### ISSUE TITLE
`struct is not supported in `trait`s or `impl`s` or: where can `utoipa::path()` be used?

### ISSUE THREAD

**Issue by Swandog** [juhaku/utoipa#136]
I have an existing `warp` project that I want to add an OpenAPI spec to. But I had a problem: the handler functions for filters are in traits, so when I tried to use the `utoipa::path` macro in front of the handler function, I got this compilation error:

```
error: struct is not supported in `trait`s or `impl`s
   --> cloud_backend/src/crud.rs:465:5
    |
465 | /     #[utoipa::path(get,
466 | |         path = "dataset/",
467 | |         responses(
468 | |             (status = 200, body = String)
469 | |         )
470 | |     )]
    | |______^
    |
    = help: consider moving the struct out to a nearby module scope
    = note: this error originates in the attribute macro `utoipa::path` (in Nightly builds, run with -Z macro-backtrace for more info)

error: implementation is not supported in `trait`s or `impl`s
   --> cloud_backend/src/crud.rs:465:5
    |
465 | /     #[utoipa::path(get,
466 | |         path = "dataset/",
467 | |         responses(
468 | |             (status = 200, body = String)
469 | |         )
470 | |     )]
    | |______^
    |
    = help: consider moving the implementation out to a nearby module scope
    = note: this error originates in the attribute macro `utoipa::path` (in Nightly builds, run with -Z macro-backtrace for more info)
```

My interpretation of this is that the `utoipa::path()` macro writes a struct definition, which is then illegal because Rust does not allow those to be defined inside a trait.

Do I have any alternatives here? It would be difficult for me to rewrite the project to extract the handlers out of the trait. Would I have to write stub handlers that call the trait methods, solely for the purpose of being annotated by `utoipa`?

My ideal option would be if I could annotate Filter objects, because that's the most basic part of my code.

Alternatively--can I write these structs myself? That may be less exciting but would make this work.

I apologize if I misunderstood something; I'm somewhat new to Rust and warp, and I just found out about utoipa today.

**Comment by Swandog** [comment]
If it helps, this is a toy version of what I'm working with:

```rust
use async_trait::async_trait;
use warp::Filter;

#[async_trait]
trait Animal {
    const NAME: &'static str;
    const SOUND: &'static str;

    async fn speak_handler() -> String {
        format!("{}", Self::SOUND)
    }
}

struct Dog;

impl Animal for Dog {
    const NAME: &'static str = "dog";
    const SOUND: &'static str = "woof";
}

struct Cat;

impl Animal for Cat {
    const NAME: &'static str = "cat";
    const SOUND: &'static str = "meow";
}

fn generate_animal_endpoint<A: Animal>(
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone
where
    A: Send + Sync + 'static,
{
    warp::path("animal")
        .and(warp::path(A::NAME))
        .then(A::speak_handler)
}

#[tokio::main]
async fn main() {
    // GET /animal/dog => 200 OK with body "woof"
    let dog_filter = generate_animal_endpoint::<Dog>();

    // GET /animal/cat => 200 OK with body "meow"
    let cat_filter = generate_animal_endpoint::<Cat>();

    let filters = dog_filter.or(cat_filter);

    warp::serve(filters).run(([127, 0, 0, 1], 3030)).await;
}
```

**Comment by juhaku** [comment]
Yes, that is actually what happens. The `utoipa::path` is implemented solely for purpose of decorating free standing functions with it. 
> My interpretation of this is that the utoipa::path() macro writes a struct definition, which is then illegal because Rust does not allow those to be defined inside a trait.
>
>Do I have any alternatives here? It would be difficult for me to rewrite the project to extract the handlers out of the trait. Would I have to write stub handlers that call the trait methods, solely for the purpose of being annotated by utoipa?

Yes there are 2 ways to get around this and one is as you mentioned to have the stub handlers outside the trait. Or to manually write the path operation. You can see some example from here: https://docs.rs/utoipa/1.0.2/utoipa/trait.Path.html
For example you could implement it manually something like this:
```rust
#[derive(OpenApi)]
    #[openapi(
        ...
        modifiers(&CustomPaths)
    )]
    struct ApiDoc;

    struct CustomPaths;

    impl Modify for CustomPaths {
        fn modify(&self, openapi: &mut utoipa::openapi::OpenApi) {
            openapi
                .paths
                .paths
                .insert(String::from("/animal/cat"), PathItem::new(PathItemType::Get, 
                        OperationBuilder::new()).build());

            openapi
                .paths
                .paths
                .insert(String::from("/animal/dog"), PathItem::new(PathItemType::Get, 
                        OperationBuilder::new()).build());
        }
    }
```
And of course since it is just a Rust you can organize the code bit better and reuse functions as usual with the manual implementation. 

> I apologize if I misunderstood something; I'm somewhat new to Rust and warp, and I just found out about utoipa today.

No worries, I'm somewhat new to Rust myself too :joy:

**Comment by Swandog** [comment]
Ah, neat! I had seen the example in `Path`, but I think what I was missing was the `Modify` part.

I'll give that a shot and let you know how it goes! Thanks!

**Comment by juhaku** [comment]
Yeah.. Indeed there is no complete example of how to add the path operation using manual implementation. This is indeed something to improve about the docs. And have to admit that the way to add  path items are not super ergonomic at the moment and there is room for an improvement. But anywhow hope that it helps you further.

**Comment by Swandog** [comment]
I wanted to update you--I went on vacation shortly after my original query.

I got finished using this method just recently. It works! But there is a LOT of code to write.

I will have to see how much time I have but I'd like to help this project. But thanks for all you have already!

**Comment by juhaku** [comment]
Thanks for update, I hope you had a good vacation. Yes indeed the manual implementation is really boiler plate heavy. Sure thing all the help is welcome. :)

**Comment by Swandog** [comment]
IDK what the right resolution is here, but I'm going to close it as "resolved" since you answered my question.

**Comment by juhaku** [comment]
Yes probably closed is the correct one to for this. This is quite unique case for declaring endpoints. And does not actually come across with other frameworks that much. It might be worth pondering whether there is something that could be done to improve the handlers in traits support but it won't be a priority. 

So far I have no actual solution for them and I only hope that if support for them were to be implemented it would't be a massive rewriting process to the core of the library. But anyway how it has been good and useful library in other aspects. And of course contributions are welcome to improve the library. :)
